File: algorithms/medium/rearrange_array_elements_by_sign.ex
defmodule Solution do
  @spec rearrange_array(nums :: [integer]) :: [integer]
  def rearrange_array(nums) do
    {p, n} = Enum.split_with(nums, fn x -> x > 0 end)
    aux(p, n)
  end

  def aux([], []), do: []

  def aux([p | ps], [n | ns]) do
    [p, n | aux(ps, ns)]
  end
end

File: algorithms/medium/rearrange_array_elements_by_sign.rs
impl Solution {
    pub fn rearrange_array(nums: Vec<i32>) -> Vec<i32> {
        let mut ans = vec![0; nums.len()];
        let mut i = 0;
        let mut j = 1;
        for &n in &nums {
            if n > 0 {
                ans[i] = n;
                i += 2;
            } else {
                ans[j] = n;
                j += 2;
            }
        }
        ans
    }
}
